crates/neon/src/types_impl/extract/error.rs

```diff
@@ -0,0 +1,178 @@
+        matches!(self.kind, Some(ErrorKind::RangeError))
+    }
+
+    /// Check if error is a `TypeError`
+    pub fn is_type_error(&self) -> bool {
+        matches!(self.kind, Some(ErrorKind::TypeError))
+    }
+
+    /// Get a reference to the underlying `cause`
+    pub fn cause(&self) -> &BoxError {
+        &self.cause
+    }
+
+    /// Extract the `std::error::Error` cause
+    pub fn into_cause(self) -> BoxError {
+        self.cause
+    }
+
+    fn create<E>(kind: ErrorKind, cause: E) -> Self
+    where
+        E: Into<BoxError>,
+    {
+        Self {
+            cause: cause.into(),
+            kind: Some(kind),
+        }
+    }
+}
+
+// Blanket impl allow for ergonomic `?` error handling from typical error types (including `anyhow`)
+impl<E> From<E> for Error
+where
+    E: Into<BoxError>,
+{
+    fn from(cause: E) -> Self {
+        Self::new(cause)
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}: {}", self.kind, self.cause)
+    }
+}
+
+// `TryFromJs` followed by `TryIntoJs` is *lossy*. We cannot guarantee the same type
+// will be created.
+impl<'cx> TryIntoJs<'cx> for Error {
+    type Value = JsError;
+
+    fn try_into_js<C>(self, cx: &mut C) -> JsResult<'cx, Self::Value>
+    where
+        C: Context<'cx>,
+    {
+        let message = self.cause.to_string();
+
+        match self.kind {
+            Some(ErrorKind::RangeError) => cx.range_error(message),
+            Some(ErrorKind::TypeError) => cx.type_error(message),
+            _ => cx.error(message),
+        }
+    }
+}
+
+impl<'cx> TryFromJs<'cx> for Error {
+    type Error = TypeExpected<JsError>;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        let err = match v.downcast::<JsError, _>(cx) {
+            Ok(err) => err,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(Self {
+            cause: get_message(cx, err)?.into(),
+            kind: get_kind(cx, err)?,
+        }))
+    }
+
+    fn from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Self>
+    where
+        C: Context<'cx>,
+    {
+        Self::try_from_js(cx, v)?.or_throw(cx)
+    }
+}
+
+fn get_message<'cx, C>(cx: &mut C, err: Handle<JsError>) -> NeonResult<String>
+where
+    C: Context<'cx>,
+{
+    let message = err
+        .get_value(cx, "message")?
+        .downcast::<JsString, _>(cx)
+        .map(|v| v.value(cx))
+        .unwrap_or_default();
+
+    Ok(message)
+}
+
+fn get_kind<'cx, C>(cx: &mut C, err: Handle<JsError>) -> NeonResult<Option<ErrorKind>>
+where
+    C: Context<'cx>,
+{
+    let name = match err.get_value(cx, "name")?.downcast::<JsString, _>(cx) {
+        Ok(v) => v.value(cx),
+        Err(_) => return Ok(None),
+    };
+
+    Ok(Some(match name.as_str() {
+        "TypeError" => ErrorKind::TypeError,
```

Comment:
> Would you consider a change in the round-tripping behavior a breaking change?

Can you say more? I don't think I understand the question.

Reply:
Consider the following:

1. Call `new SyntaxError("oh, no!")` in JavaScript
2. Pass the error to Neon
3. Neon extracts the error into `neon::types::extract::Error` with it's `TryFromJs` implementation
4. Neon converts it *back* into a JavaScript error with `TryIntoJs`

This error started as a `SyntaxError`, but ended as an `Error` because Neon doesn't have a representation of a `SyntaxError`.

In the future, we add `Error::syntax_error(..)` and `Error` now understands `SyntaxError`. Behavior has changed. The end type is now a `SyntaxError` instead of an `Error`.

Is this a breaking change? I'm inclined to say *no* because it would be a very niche use case. I can't actually think of a use case to extract a JavaScript error. I only added it for symmetry in the implementation (`TryFromJs` implementation to match `TryIntoJs`).

If you feel this *would* be a breaking change, my preference is to completely remove the `TryFromJs` implementation.

Reply:
I'm not super worried about that case, but I also think we can wait until we have clearer evidence of a need for `extract::Error` and just remove it for now.

Reply:
As an example, [ECMAScript 2021 introduced `AggregateError`](https://tc39.es/ecma262/2021/#sec-intro) as a new error type, so if and when we do ship an error extractor, we'll have to expect that new error types could come along any time.